tests/language-server-simulator/src/main/java/org/ballerinalang/langserver/simulator/EditorTab.java

```diff
@@ -0,0 +1,238 @@
+        } catch (IOException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Simulates a user typing the provided content in the editor. Content is typed character by character similarly to
+     * how a user does it.
+     *
+     * @param content Text content to be typed in the editor.
+     */
+    public void type(String content) {
+        int missCount = 0;
+        for (int i = 0; i < content.length(); i++) {
+            String typedChar = Character.toString(content.charAt(i));
+
+            int startOffset = textDocument.textPositionFrom(LinePosition.from(cursor.getLine(), cursor.getCharacter()));
+            TextEdit edit = TextEdit.from(TextRange.from(startOffset, 0), typedChar);
+            TextDocumentChange change = TextDocumentChange.from(new TextEdit[]{edit});
+            textDocument = textDocument.apply(change);
+
+            LinePosition newLinePos = textDocument.linePositionFrom(startOffset + 1);
+            try {
+                TestUtil.didChangeDocument(this.endpoint, this.filePath, textDocument.toString());
+            } catch (Throwable t) {
+                logger.error("Caught error in didChange", t);
+            }
+            cursor(newLinePos.line(), newLinePos.offset());
+            // logger.info("Added char: {} and cursor advanced to: {}", typedChar, newLinePos);
+
+            if (i % 10 == 0) {
+                float completionPercentage = ((float) i / (float) content.length()) * 100;
+                writer.printf("%.1f%%\r", completionPercentage);
+            }
+
+            // Get completions in the background
+            if (i % 3 == 0) {
+                CompletableFuture.runAsync(this::completions);
+                CompletableFuture.runAsync(this::codeActions);
+            }
+
+            if (!isDocumentInSync()) {
+                missCount++;
+            }
+
+            try {
+                Thread.sleep(100 + random.nextInt(300));
+            } catch (InterruptedException e) {
+                logger.error("Interrupted", e);
+                break;
+            }
+        }
+        logger.info("Typed provided content in file: {} -> \n{}",
+                filePath, content.substring(0, Math.min(20, content.length())));
+        logger.info("Typed {} characters with {} out of sync scenarios", content.length(), missCount);
+
+        while (!isDocumentInSync()) {
+            logger.info("Document out of sync. Waiting 30 seconds and syncing...");
+            try {
+                Thread.sleep(30 * 1000);
+            } catch (InterruptedException e) {
+                break;
+            }
+            TestUtil.didChangeDocument(this.endpoint, this.filePath, textDocument.toString());
+        }
+    }
+
+    /**
+     * Check if the document in this instance is similar to that is in the language server.
+     *
+     * @return
+     */
+    private boolean isDocumentInSync() {
```

Comment:
This method is always inverted. Shall we send the negation?

Reply:
Done